File: laboratorioalgoritmi-2015-16/Esercizio1/Insertionsort/InsertionSort.java
public class InsertionSort{

	public static void swap(int [] array, int j, int i){
		int k = array[j];
		array[j] = array[i];
		array[i] = k;

	}

	public static void insertionSort(int[] array){
		for(int i = 1; i < array.length; i++){
			int j = i-1;
        	for(; j >= 0; j--){
        		if(array[i] < array[j]){
        			swap(array, j, i);
        			i--;
        		}else
        			break;
        	}
    	}
    	for(int i = 0; i < array.length; i++){
    	System.out.print(array[i] + ", ");

    	}
    	System.out.println();
	}


}

File: laboratorioalgoritmi-2015-16/Esercizio1/Insertion_Sort/insertion_sort.c
#include "insertion_sort.h"
#include <stdio.h>

void swap(void** e1, void** e2) {
  void* tmp = *e1;
  *e1 = *e2;
  *e2 = tmp;
}	

void insertion_sort(void** array, int size, CompareFunc compare) {
  int i, j;
  
 	for(i = 1; i < size; ++i){
  	j = i - 1;
  	while(j >= 0 && (compare(array[j], array[i]) > 0)){
  		swap(&array[i], &array[j]);
  		i = j;
  		j--;
  	}
  }
}